dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,254 @@
+                output[output_column] = ""
+            else:
+                for output_row in output:
+                    output_row[output_column] = ""
+        try:
+            response = (
+                self.function(row=row, **function_kwargs) if row else self.function(batch=batch, **function_kwargs)
+            )
+            if row:
+                output[output_column_names.response] = response
+            else:
+                output = self.batch_response_parser(
+                    batch=batch, response=response, output_column_names=output_column_names
+                )
+                errors = [
+                    row[output_column_names.error_message] for row in output if row[output_column_names.error_message]
+                ]
+                if errors:
+                    raise BatchError(str(errors))
+        except self.exceptions_to_catch + (BatchError,) as error:
+            if self.error_handling == ErrorHandling.FAIL:
+                raise error
+            logging.warning(
+                f"Function {self.function.__name__} failed on: {row if row else batch} because of error: {error}"
+            )
+            error_type = str(type(error).__qualname__)
+            module = inspect.getmodule(error)
+            if module:
+                error_type = f"{module.__name__}.{error_type}"
+            if row:
+                output[output_column_names.error_message] = str(error)
+                output[output_column_names.error_type] = error_type
+                output[output_column_names.error_raw] = str(error.args)
+            else:
+                for output_row in output:
+                    output_row[output_column_names.error_message] = str(error)
+                    output_row[output_column_names.error_type] = error_type
+                    output_row[output_column_names.error_raw] = str(error.args)
+        return output
+
+    def _convert_results_to_df(
+        self, df: pd.DataFrame, results: List[Dict], output_column_names: NamedTuple,
+    ) -> pd.DataFrame:
+        """Combine results from the function with the input dataframe"""
+        output_schema = {**{column_name: str for column_name in output_column_names}, **dict(df.dtypes)}
+        output_df = (
+            pd.DataFrame.from_records(results)
+            .reindex(columns=list(df.columns) + list(output_column_names))
+            .astype(output_schema)
+        )
+        if not self.verbose:
+            output_df.drop(labels=output_column_names.error_raw, axis=1, inplace=True)
+        if self.error_handling == ErrorHandling.FAIL:
+            error_columns = [
+                output_column_names.error_message,
+                output_column_names.error_type,
+                output_column_names.error_raw,
+            ]
+            output_df.drop(labels=error_columns, axis=1, inplace=True, errors="ignore")
+        return output_df
+
+    def run(self, df: pd.DataFrame, **function_kwargs,) -> pd.DataFrame:
```

Comment:
Interesting. I have tried to keep it all in one class so that the developer user can "run" in 3 lines of code.

```
from dkulib.parallelizer import DataFrameParallelizer
parallelizer = DataFrameParallelizer(function=call_api, exceptions_to_catch=api_exceptions)
output_df = parallelizer.run(input_df, api_function_param="invalid_integer")
```

If we add a second Runner class, how would the snippet look like?

Reply:
Point taken. The refactoring may take some time. Should I address it here or in another PR?

Reply:
Hi @arnaudde,

According to our last conversation, I created a dedicated CH story: https://app.clubhouse.io/dataiku/story/65582/parallelizer-separate-runner-class-to-simplify-code.

I can discuss with Mehdi and Niklas for them to implement this refactoring story after we merge this PR.

Cheers,

Alex

Reply:
@arnaudde I have discussed this story with @Muennighoff. He raised a good question: is this new refactoring still necessary given that [ch65583](https://app.clubhouse.io/dataiku/story/65583/parallelizer-hold-data-in-lists-for-both-batch-and-row-by-row-case) has made the code much shorter.